plugins/artifacts/artifacts.go

```diff
@@ -31,40 +41,101 @@ type pluginProvider struct {
 
 type plugin struct {
 	plugins.PluginBase
+	environment *runtime.Environment
+	privateKey  *openpgp.Entity // nil, if COT is disabled
+	config      config
 }
 
 type taskPlugin struct {
 	plugins.TaskPluginBase
-	context     *runtime.TaskContext
-	artifacts   []artifact
-	monitor     runtime.Monitor
-	failed      atomics.Bool                    // If true, Stopped() returns false
-	mErrors     sync.Mutex                      // Guards errors
-	errors      []runtime.MalformedPayloadError // errors to be returned from Stopped()
-	nonFatalErr atomics.Bool                    // if true, Stopped() returns non-fatal error
-	fatalErr    atomics.Bool                    // if true, Stopped() returns fatal error
+	plugin       *plugin
+	context      *runtime.TaskContext
+	artifacts    []artifact
+	createCOT    bool
+	certifiedLog bool
+	uploaded     map[string][]byte // Map from artifact to sha256 hash
+	mUploaded    sync.Mutex
+	monitor      runtime.Monitor
+	failed       atomics.Bool                    // If true, Stopped() returns false
+	mErrors      sync.Mutex                      // Guards errors
+	errors       []runtime.MalformedPayloadError // errors to be returned from Stopped()
+	nonFatalErr  atomics.Bool                    // if true, Stopped() returns non-fatal error
+	fatalErr     atomics.Bool                    // if true, Stopped() returns fatal error
 }
 
 func init() {
 	plugins.Register("artifacts", pluginProvider{})
 }
 
-func (pluginProvider) NewPlugin(plugins.PluginOptions) (plugins.Plugin, error) {
-	return &plugin{}, nil
+func (pluginProvider) ConfigSchema() schematypes.Schema {
+	return configSchema
+}
+
+func (pluginProvider) NewPlugin(options plugins.PluginOptions) (plugins.Plugin, error) {
+	var c config
+	schematypes.MustValidateAndMap(configSchema, options.Config, &c)
+
+	var key *openpgp.Entity
+	if c.PrivateKey != "" {
+		keyring, err := openpgp.ReadArmoredKeyRing(bytes.NewBufferString(c.PrivateKey))
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed to load private key")
```

Comment:
nit: add the `err` object to error strings. I can't count how many hours I lost because a log didn't have the original error message.

Reply:
It's in there, the `errors.Wrap(err, ....)` does the right thing :)
In fact it shows some sort of trace, which is really nice...

Checkout: https://godoc.org/github.com/pkg/errors#example-Wrap--Extended

Reply:
Oh, my bad, pretend I didn't make this comment :)